src/rhsmlib/facts/cloud_facts.py

```diff
@@ -47,57 +40,104 @@ def __init__(self, arch=None, prefix=None, testing=None, collected_hw_info=None)
             collected_hw_info=collected_hw_info
         )
 
-        self.hardware_methods = [
-            self.get_aws_instance_id
-        ]
-
-    # https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/identify_ec2_instances.html
-    def get_aws_instance_id(self):
+        self.hardware_methods = []
+
+        # Try to detect cloud provider
+        self.cloud_provider = get_cloud_provider(self._collected_hw_info)
+        if self.cloud_provider is not None:
+            # Create dispatcher for supported cloud providers
+            cloud_provider_dispatcher = {
+                "aws": self.get_aws_facts,
+                "azure": self.get_azure_facts,
+                "gcp": self.get_gcp_facts
+            }
+            # Set method according detected cloud provider
+            if self.cloud_provider.CLOUD_PROVIDER_ID in cloud_provider_dispatcher:
+                self.hardware_methods = [
+                    cloud_provider_dispatcher[self.cloud_provider.CLOUD_PROVIDER_ID]
+                ]
```

Comment:
This is _literally_ a class method dispatch :)

Please make it as a proper `BaseCloudProvider.get_facts()` method, to be reimplemented in provider subclasses. This way, there is no need for this manual dispatch here, and `CloudFactsCollector` is finally free of code specific for cloud providers.

Reply:
I don't see how this would impact the split of this code in its own package. My idea is to simply hide `CloudFactsCollector` and all its subclasses, and make `CloudFactsCollector` the only interface to query everything, both facts and metadata. Hence, adding a `BaseCloudProvider.get_facts()` would be only an implementation detail that can be changed anytime.